x-pack/plugins/apm/server/routes/agent_keys/create_agent_key.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import Boom from '@hapi/boom';
+import { ApmPluginRequestHandlerContext } from '../typings';
+import { CreateApiKeyResponse } from '../../../common/agent_key_types';
+
+const enum PrivilegeType {
+  SOURCEMAP = 'sourcemap:write',
+  EVENT = 'event:write',
+  AGENT_CONFIG = 'config_agent:read',
+}
+
+interface SecurityHasPrivilegesResponse {
+  application: {
+    apm: {
+      '-': {
+        [PrivilegeType.SOURCEMAP]: boolean;
+        [PrivilegeType.EVENT]: boolean;
+        [PrivilegeType.AGENT_CONFIG]: boolean;
+      };
+    };
+  };
+  has_all_requested: boolean;
+  username: string;
+}
+
+export async function createAgentKey({
+  context,
+  requestBody,
+}: {
+  context: ApmPluginRequestHandlerContext;
+  requestBody: {
+    name: string;
+    sourcemap?: boolean;
+    event?: boolean;
+    agentConfig?: boolean;
+  };
+}) {
+  // Elasticsearch will allow a user without the right apm privileges to create API keys, but the keys won't validate
+  // check first whether the user has the right privileges, and bail out early if not
+  const {
+    body: { application, username, has_all_requested: hasRequiredPrivileges },
+  } = await context.core.elasticsearch.client.asCurrentUser.security.hasPrivileges<SecurityHasPrivilegesResponse>(
+    {
+      body: {
+        application: [
+          {
+            application: 'apm',
+            privileges: [
+              PrivilegeType.SOURCEMAP,
+              PrivilegeType.EVENT,
+              PrivilegeType.AGENT_CONFIG,
+            ],
+            resources: ['-'],
+          },
+        ],
+      },
+    }
+  );
+
+  if (!hasRequiredPrivileges) {
+    const missingPrivileges = Object.entries(application.apm['-'])
+      .filter((x) => !x[1])
+      .map((x) => x[0])
+      .join(', ');
+    const error = `${username} is missing the following requested privilege(s): ${missingPrivileges}.\
```

Comment:
I believe it makes sense to show this error in the UI as this requirement doesn't seem to be documented. Is it ok if we display whichever error is thrown from this route or hardcode the text in the UI as well?

Reply:
@cauemarcondes @sqren what do you think?